pynumdiff/optimize/_optimize.py

```diff
@@ -0,0 +1,186 @@
+                    structured as :code:`{param1:[values], param2:[values], param3:value, ...}`. The search space
+                    is the Cartesian product. If left None, a default search space of initial values is used.
+    :param np.array[float] dxdt_truth: actual time series of the derivative of x, if known
+    :param float tvgamma: regularization value used to select for parameters that yield a smooth derivative.
+                    Larger value results in a smoother derivative
+    :param int padding: number of time steps to ignore at the beginning and end of the time series in the
+                    optimization. Larger value causes the optimization to emphasize the accuracy of dxdt in the
+                    middle of the time series
+    :param str metric: either :code:`'rmse'` or :code:`'error_correlation'`, only applies if :code:`dxdt_truth`
+                    is not None, see _objective_function
+    :param str opt_method: Optimization technique used by :code:`scipy.minimize`, the workhorse
+    :param int maxiter: passed down to :code:`scipy.minimize`, maximum iterations
+
+    :return: tuple[dict, float] of\n
+            - **opt_params** -- best parameter settings for the differentation method
+            - **opt_value** -- lowest value found for objective function
+    """
+    if metric not in ['rmse','error_correlation']:
+        raise ValueError('`metric` should either be `rmse` or `error_correlation`.')
+    if metric == 'error_correlation' and dxdt_truth is None:
+        raise ValueError('`metric` can only be `error_correlation` if `dxdt_truth` is given.')
+
+    params, bounds = method_params_and_bounds[func]
+    params.update(search_space) # for things not given, use defaults
+
+    # No need to optimize over singletons, just pass them through
+    singleton_params = {k:v for k,v in params.items() if not isinstance(v, list)}
+
+    # The search space is the Cartesian product of all dimensions where multiple options are given
+    search_space_types = {k:type(v[0]) for k,v in params.items() if isinstance(v, list)} # for converting back and forth from point
+    if any(v not in [float, int, bool] for v in search_space_types.values()):
+        raise ValueError("Optimization over categorical strings currently not supported")
+    # If excluding string type, I can just cast ints and bools to floats, and we're good to go
+    cartesian_product = product(*[np.array(params[k]).astype(float) for k in search_space_types])
+    
+    bounds = [bounds[k] if k in bounds else # pass these to minimize(). It should respect them.
+             (0, 1) if v == bool else
+             None for k,v in search_space_types.items()]
+
+    # wrap the objective and scipy.optimize.minimize because the objective and options are always the same
+    _obj_fun = partial(_objective_function, func=func, x=x, dt=dt, singleton_params=singleton_params,
```

Comment:
I'm trying to keep down the amount of information flying around. The use of `partial` and the comment make clear that essentially a bunch of stuff that flows in to the `optimize` function stays constant throughout the process, so we only need to focus on the `point` (and translation to and from the full parameter dictionaries).